### ISSUE TITLE
Force int8 execution and algorithm selector

### ISSUE THREAD

**Issue by yuanjiechen** [NVIDIA/TensorRT#3074]
## Description

I use QAT (pytorch-quantization) to build an explicitly quantized yolov5 model, and the target accuracy is int8. I took the route of torch->onnx->trt, but no matter how optimized, most layers still work in fp16 when fp16 is enabled, and disable In the case of fp16 it will fall back to fp32, only a few layers work in int8.
Considering that it may be a problem of layer fusion, I replaced SiLU with ReLU and retrained to obtain the acceleration of layer fusion, but it is still impossible to realize that most layers run on int8. (PTQ implicit quantization of int8 has poor performance)

i have two questions
1. How to force the layer to run under int8, which api should be used to do this under explicit quantization? (I have tried other small models, which can run under int8 by default)
2. For the algorithm selector in the document, how should I know what each algorithm does, and then choose the optimization method for each layer?

Thanks

## Environment

**TensorRT Version**: 8.5.1.7

**NVIDIA GPU**: RTX 3090

**NVIDIA Driver Version**: 515.48

**CUDA Version**: 11.6.1

**CUDNN Version**: 8.9.1

Operating System: ubuntu 20.02

Python Version (if applicable): 3.9.12

Tensorflow Version (if applicable): /

PyTorch Version (if applicable): 2.0.1

Baremetal or Container (if so, version): /

## Relevant Files

**Model link**:

## Steps To Reproduce

**Commands or scripts**:

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**:

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`):

**Comment by zerollzeng** [comment]
@ttyio ^ ^

**Comment by ttyio** [comment]
Hi @yuanjiechen 
1. Here is the document on how explicit quantization works in TRT https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#work-with-qat-networks, you need follow this doc the place the Q/DQ in the correct position for some of the network.
2. Each id in the algorithm selector is black box. let's rely on the TRT default optimizer to select the fastest kernel. Algorithm selector is used when you either want a determined engine or you want to blacklist some kernels.

**Comment by yuanjiechen** [comment]
> Hi @yuanjiechen
> 
> 1. Here is the document on how explicit quantization works in TRT https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#work-with-qat-networks, you need follow this doc the place the Q/DQ in the correct position for some of the network.
> 2. Each id in the algorithm selector is black box. let's rely on the TRT default optimizer to select the fastest kernel. Algorithm selector is used when you either want a determined engine or you want to blacklist some kernels.

Hi @ttyio:
Thanks for your reply!
I follow the document and put Q/DQ node in network by pytorch-quantization, and enable the build flag. 
But most of layers still work under fp16 not int8, and I tried to put Q/DQ before/after relu the problem still exists (just try, I know the document about layer fusion said: don't put Q/DQ here).
Any other ideas to let the model run under int8? (the layer fusion is working normal when I replace the SiLU to ReLU).

Thanks

**Comment by ttyio** [comment]
8.5 might only support fusion depthwise conv + siLU into int8 kernel. @yuanjiechen  Have you take a try with the 8.6GA? thanks!

**Comment by yuanjiechen** [comment]
Hi @ttyio:
Thanks for your reply!
I have updated the trt to 8.6.1.6 and build the engine again. 
I enable the int8 flag but the most of layers still run under fp32. 
Layer fusion works normally.
Here is my onnx model and profile result.
Could you please provide more advice?
https://drive.google.com/file/d/1HogKK2y-Qw_M83yziMIFcLlEUcoASIrn/view?usp=sharing
[profile.txt](https://github.com/NVIDIA/TensorRT/files/11903468/profile.txt)
[layerinfo.txt](https://github.com/NVIDIA/TensorRT/files/11903469/layerinfo.txt)
ps: layerinfo and profile are both json files. 

Thanks !

**Comment by yuanjiechen** [comment]
Oh I find where the bug is: it's from yolo save and load post-process.
I'm not sure why these function remove 90% qdq node.
Anyway, I can debug from here and thanks your time!

**Comment by yuanjiechen** [comment]
Hi @zerollzeng :
After I update trt to 8.6, I also reinstall cuda to 11.7.1 with 515 dirver, cudnn 8.9.1.
The pytorch training,  trt mnist_sample, engine build is working normal, but inference with the following error: 
[07/03/2023-14:22:52] [TRT] [W] CUDA initialization failure with error: 35
Segmentation fault (core dumped)
The inference code used in trt 8.4 and 8.6 is same. 
The code died in this operation:
            with open(w, 'rb') as f, trt.Runtime(logger) as runtime:
                model = runtime.deserialize_cuda_engine(f.read())
Any other ideas to solve this?
Thanks

**Comment by ttyio** [comment]
@yuanjiechen , `35` means `cudaErrorInsufficientDriver` according to https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038f5e52d1774934b77ba55d2aa2c063067

Could you upgrade your driver? thanks!

**Comment by yuanjiechen** [comment]
Hi @ttyio :
I have read the document but the driver is follow the cuda installation. 
If I install cuda 11.7.1 (for trt 8.6), the driver is 515, and cuda 11.8 (for trt 8.6), the driver will auto updated to 520.
If I install driver 520 use apt, the cuda 11.7.1 will be auto removed. 
But when I rollback trt to 8.4, the engine build and inference normally. (cuda still in 11.8 and cudnn 8.9.1, driver 520)
The pytorch and other components working normally. 

Another question, I fix all qdq nodes, and finetune the model. 
I think the qdq node simulate the quant and clip error, the build engine need get almost same performance as the pytorch 
fake quant model, am I right?

THanks

**Comment by ttyio** [comment]
@yuanjiechen , let's use 525 driver since we use CUDA 12.x in 8.6, see https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#id4

For the accuracy, let's upgrade driver and test again with TRT8.6, thanks!

**Comment by yuanjiechen** [comment]
Thanks @ttyio :
cuda 12.1.1 with trt 8.6 works and all performance are well. 
Layer fusion on sigmoid also work!

Thanks for your time!
